### PR TITLE
Extend paygas JSON tests

### PR DESCRIPTION
**Pull request by jannikluhn**
- [x] Test the value PAYGAS pushes on to the stack (denoting success or failure)
- [x] Test that the REVERT opcode does not revert gas payment

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by jannikluhn** [comment]
I'm reasonably confident that the tests are correct, but it would be nice nevertheless to have another pair of eyes to look over them. Preferably, @NIC619 could review as he's written the implementation of the opcode.

Looking at the JSON files directly probably doesn't make a whole lot of sense, but https://github.com/jannikluhn/py-evm/blob/0812496bab24c91a9b9b1e12bdb814a080958809/tests/fillers/vm_fillers/paygas_fillers.py is supposed to be somewhat readable. To check correctness, one should look at the Vyper LLL code, the pre state and the post state changes. Usually the balance of the `address` account is reduced by a little as it had to pay for gas. The return code of paygas is usually stored in the storage, which is then checked with `paygas_successful` or `paygas_failed` in the post state. Also, the tests store something to check if the state is reverted (if it isn't supposed to be reverted, `storage_updated` is put into the post state).

**Comment by pipermerriam** [comment]
I re-directed this PR to be towards `master` but it needs to be rebased now.

File: tests/fillers/vm_fillers/paygas_fillers.py [hunk 1]
```diff
@@ -32,15 +25,18 @@
     1: 1,
 })
 
-broke_caller = (caller, "balance", 0)
-solvent_caller = (caller, "balance", 1000000)
+paygas_successful = (address, "storage", 2, 2)  # return code + 1
+paygas_failed = (address, "storage", 2, 1)  # return code + 1
```
Comment (NIC619) [inline tests/fillers/vm_fillers/paygas_fillers.py]: In `paygas_in_call_test` test case, there doesn't seem to be an `ADD` operation.

File: tests/fillers/vm_fillers/paygas_fillers.py [hunk 2]
```diff
@@ -327,11 +337,41 @@
...
+    expect(
+        networks=["Sharding"],
+        transaction={
+            "to": address,
+            "accessList": [
+                [address, b""],
+            ],
+        },
+        post_state=[
+            (address, "balance", 18951),
```
Comment (NIC619) [inline tests/fillers/vm_fillers/paygas_fillers.py]: Should the balance in post state be `pre_state_balance - transaction_gas_limit * gas_price`? As all gas will be confiscated.

File: tests/fillers/vm_fillers/paygas_fillers.py (hunk 1 above)
Comment (jannikluhn) [reply tests/fillers/vm_fillers/paygas_fillers.py]: It is, but it's hidden in the `paygas_contract`. See https://github.com/jannikluhn/py-evm/blob/0812496bab24c91a9b9b1e12bdb814a080958809/tests/fillers/vm_fillers/paygas_fillers.py#L37

File: tests/fillers/vm_fillers/paygas_fillers.py (hunk 2 above)
Comment (jannikluhn) [reply tests/fillers/vm_fillers/paygas_fillers.py]: No, here state is reverted with the REVERT opcode which doesn't consume all gas. See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-140.md

File: tests/fillers/vm_fillers/paygas_fillers.py (hunk 1 above)
Comment (NIC619) [reply tests/fillers/vm_fillers/paygas_fillers.py]: But aren't both `paygas_successful` and `paygas_failed` the cases for `address`(which is generated by `generate_random_address()`)?
`paygas_successful ` is the case for `address` but shouldn't `paygas_failed` be the case for `paygas_contract_address`?

File: tests/fillers/vm_fillers/paygas_fillers.py (hunk 2 above)
Comment (NIC619) [reply tests/fillers/vm_fillers/paygas_fillers.py]: Yep you are right. I mistake the `assert` here for the one in Solidity(`asset` there will take away all the gas).

File: tests/fillers/vm_fillers/paygas_fillers.py (hunk 1 above)
Comment (jannikluhn) [reply tests/fillers/vm_fillers/paygas_fillers.py]: Ah, now I get what you mean. This is what's going on: The paygas contract returns the return code (+ 1), but the "main" contract (at `address`) stores it in its storage. So to check for success/failure, one has to check the storage at `address`, even if the actual call was done at `paygas_contract_address`. See https://github.com/jannikluhn/py-evm/blob/0812496bab24c91a9b9b1e12bdb814a080958809/tests/fillers/vm_fillers/paygas_fillers.py#L274

Comment (NIC619) [reply tests/fillers/vm_fillers/paygas_fillers.py]: Ah I see. I didn't see that in `paygas_contract`, result of `["ADD", 1, ["PAYGAS", 1]]` is store in memory and then returned.

**Pull request merged**